Fix bugs in AMR TruncationError criterion

wthrowe opened the pull request:
## Proposed changes

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

wthrowe committed:
Remove None options from AMR TruncationError

Setting either value to None completely prevented decreasing
resolution.

wthrowe committed:
Fix combining absolute and relative AMR tolerances

Previously tried to satisfy both tolerances instead of just the less
restrictive.

nilsdeppe commented on src/ParallelAlgorithms/Amr/Criteria/TruncationError.cpp:
```diff
@@ -64,11 +61,8 @@ void max_over_components(
           umax * rel_truncation_error_coarsened;
-      if ((target_rel_truncation_error.has_value() and
-           rel_truncation_error_coarsened <=
-               target_rel_truncation_error.value()) and
```
I don't generally disagree with people needing to set both, but just for a record of it: would the but be fixed by doing `value_or(std::numeric_limits<double>::inf()`?

wthrowe replied:
Yes.  Infinity before the second commit changed to zero after would also have worked.

nilsdeppe commented on src/ParallelAlgorithms/Amr/Criteria/TruncationError.cpp:
```diff
@@ -61,7 +61,7 @@ void max_over_components(
           umax * rel_truncation_error_coarsened;
-      if (rel_truncation_error_coarsened <= target_rel_truncation_error and
+      if (rel_truncation_error_coarsened <= target_rel_truncation_error or
           abs_truncation_error_coarsened <= target_abs_truncation_error) {
```
Could we recast this and the above check into something like:
```
err = abs(error)/(atol + rtol * abs(u))
```
I ask because I find that much easier to understand than the logic here, and it nicely maps onto 17.2.7-17.2.9 in Numerical Recipes. I think this does effectively that, if I understand correctly.

Up to you, just an idea on how to eliminate the logic, because I think the logic is subtle and requires more thought.

wthrowe commented:
@nilsvu has said he may disagree with this change, and has asked that we wait until he has a change to look at it next week.  (And has pointed out that the docs need updating.)

nilsvu commented:
Thanks @wthrowe. No I don't disagree with the change, you can go ahead. When I wrote this logic I did intend to have the criterion satisfy _both_ rel and abs target if both are specified, with the idea that you want to stay below those targets if you set them. This is also what the option helpstrings say, so I don't consider this a bugfix. You're changing the behavior to satisfy _either_ the rel or abs target. If you think that makes more sense then go ahead with the change, but also change the docs.

nilsdeppe commented:
Okay, so my (rather strong) preference is to use
$\mathcal{E}=\frac{|\Delta|}{\epsilon_{\mathrm{abs}}+|u|\epsilon_{\mathrm{rel}}}$.
Then we require $\mathcal{E}<1$ to not refine, and the same for the "one fewer point" in order to coarsen. The goal is that you smoothly transition to "not caring" about a variable whose value is too small if you set an absolute & relative tolerance. If you set an absolute tolerance of $0$ then you only use the relative tolerance, and same with setting a relative tolerance of $0$ to only care about the absolute value. I believe this is what the time steppers already do.

nilsvu commented:
Yep I agree that it makes more sense to transition between rel and abs tol for this criterion, rather than satisfy both.

wthrowe committed:
Stop returning log from relative_truncation_error

There are two overloads, one of which returned the error while the
other returned -log10 of the error.  This was very confusing, and all
actual uses immediately undid the logarithm.

wthrowe commented:
Added a fixup and a new commit changing the interface to the power monitor error because it was confusing.

nilsdeppe reviewed (commented):
LGTM! Please squash
